src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1242,17 +1242,34 @@ public static TSource Last<TSource>(this IEnumerable<TSource> source, Func<TSour
         {
             if (source == null) throw Error.ArgumentNull("source");
             if (predicate == null) throw Error.ArgumentNull("predicate");
-            TSource result = default(TSource);
-            bool found = false;
-            foreach (TSource element in source)
+            IList<TSource> list = source as IList<TSource>;
+            if (list != null)
             {
-                if (predicate(element))
+                for(var i = list.Count - 1; i >= 0; --i)
```

Comment:
Funny, I managed to subdue my normally var-heavy style, and then it crept in where I'd normally use int. 
On a phone now, but will get this soon.